### ISSUE TITLE
Install from CPAN fails on Debian 11

### ISSUE THREAD

**Issue by yrutschle** [cnangel/Conf-Libconfig#4]
Hello,

On a clean Debian 11, `cpan Conf::Libconfig` fails with:

```
x86_64-linux-gnu-gcc -c -D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -DVERSION="1.0.0" -DXS_VERSION="1.0.0" -fPIC "-I/usr/lib/x86_64-linux-gnu/perl/5.32/CORE" Libconfig.c
Libconfig.xs: In function ‘sv2addarray’:
Libconfig.xs:365:19: warning: implicit declaration of function ‘av_count’ [-Wimplicit-function-declaration]
365 | int avlen = (int)av_count(av);
| ^~~~~~~~
Libconfig.xs:391:4: error: label at end of compound statement
391 | default:
| ^~~~~~~
Libconfig.xs: In function ‘sv2addobject’:
Libconfig.xs:430:5: error: label at end of compound statement
430 | default:
| ^~~~~~~
Libconfig.xs:452:5: error: label at end of compound statement
452 | default:
| ^~~~~~~
Libconfig.xs: In function ‘set_general_value’:
Libconfig.xs:486:5: error: label at end of compound statement
486 | default:
| ^~~~~~~
Libconfig.xs:510:5: error: label at end of compound statement
510 | default:
| ^~~~~~~
Libconfig.xs:532:5: error: label at end of compound statement
532 | default:
| ^~~~~~~
Libconfig.xs:554:5: error: label at end of compound statement
554 | default:
| ^~~~~~~
Libconfig.xs: In function ‘get_general_value’:
Libconfig.xs:663:4: error: a label can only be part of a statement and a declaration is not a statement
663 | const char *val = config_setting_get_string(elem);
| ^~~~~
make: *** [Makefile:346 : Libconfig.o] Erreur 1
CNANGEL/Conf-Libconfig-1.0.0.tar.gz
/usr/bin/make -- NOT OK
```
I am wondering if it's `Perl_croak` that now somehow expands to nothing, or it always was empty and gcc now complains about it...

**Comment by yrutschle** [comment]
Never mind, that's got nothing to do with `Perl_croak`, it's the "default:" statements that are empty that are the problem. I'll prepare a PR.

**Comment by yrutschle** [comment]
Building fails with `av_count()` being unknown. I'm trying to replace it with `av_len()` or `av_top_index()` but I seem to corrupt memory, or something, as tests in 04-specread segfault.

**Comment by cnangel** [comment]
can you give me a patch for this?

**Comment by yrutschle** [comment]
I am still investigating. I find commit 2e759c86b9790fe310064c9f89fbc86096eba01a works, while 084ebb0feccac9b621c60edcd9878a9a96fa236d does not, so I guess it's linked to the change in libconfig version (Debian is on 1.5), although I haven't tried on libconfig 1.7. Unfortunately the commit that breaks is rather huge.
Is it supposed to be backwards compatible, or am I wasting my time?

**Comment by rafork** [comment]
The perl folk pointed out that av_count was introduced in Perl 5.34.0. That's a problem. Debian 11 is using Perl 5.32.1.
So the only way to fix this is to change Conf-Libconfig to not use av_count. Presumably it would be possible for you to go back to doing things the way you did before starting to use av_count. The metacpan page says that Conf::Libslack only requires perl v5.6.1. You could instead change it to require v5.34 but that would be unfortunate.

**Comment by cnangel** [comment]
Thanks for rafork, @yrutschle  Can you try it?

**Comment by rafmod** [comment]
As it stands, it should compile, but it's still the case that av_count isn't present in perl 5.32 which is what debian stable has. I think the XS code still needs to be changed so as to not require av_count. It can be defined as a macro "av_top_index(av) + 1", and doing so fixes the linking problem so the tests can all run, but it's still the case that there is a segfault in one of the test files. And I don't know if it would work with older versions of perl. I think that, if av_count is going to stay in the code, then some effort will be needed to find and fix the segfault.

**Comment by rafork** [comment]
As it stands, it should compile, but it's still the case that av_count isn't present in perl 5.32 which is what debian stable has. I think the XS code still needs to be changed so as to not require av_count. It can be defined as a macro "av_top_index(av) + 1", and doing so fixes the linking problem so the tests can all run, but it's still the case that there is a segfault in one of the test files. And I don't know if it would work with older versions of perl. I think that, if av_count is going to stay in the code, then some effort will be needed to find and fix the segfault.

**Comment by cnangel** [comment]
```bash
$make test
"/usr/bin/perl" "-Iinc" -MExtUtils::Command::MM -e 'cp_nonempty' -- Libconfig.bs blib/arch/auto/Conf/Libconfig/Libconfig.bs 644
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'inc', 'blib/lib', 'blib/arch')" t/00-load.t t/01-init.t t/02-read.t t/03-write.t t/04-specread.t t/05-add.t t/06-delete.t t/07-memory.t t/08-exception.t t/09-complex.t t/10-mutiinit.t t/11-booltest.t t/11-mutinew.t t/12-includedir.t t/boilerplate.t t/pod-coverage.t t/pod-spell.t t/pod.t
t/00-load.t ........ 1/1 # Testing Conf::Libconfig 1.0.0, Perl 5.036001, /usr/bin/perl
t/00-load.t ........ ok   
t/01-init.t ........ ok   
t/02-read.t ........ ok     
t/03-write.t ....... ok   
t/04-specread.t .... ok     
t/05-add.t ......... ok     
t/06-delete.t ...... ok     
t/07-memory.t ...... ok       
t/08-exception.t ... ok     
t/09-complex.t ..... ok   
t/10-mutiinit.t .... ok   
t/11-booltest.t .... ok     
t/11-mutinew.t ..... ok   
t/12-includedir.t .. ok    
t/boilerplate.t .... ok   
t/pod-coverage.t ... ok   
t/pod-spell.t ...... ok   
t/pod.t ............ skipped: Test::Pod 1.22 required for testing POD
All tests successful.

Test Summary Report
-------------------
t/05-add.t       (Wstat: 0 Tests: 42 Failed: 0)
  TODO passed:   40
Files=18, Tests=35543,  5 wallclock secs ( 1.58 usr  0.10 sys +  4.30 cusr  0.39 csys =  6.37 CPU)
Result: PASS

$rpm -q libconfig
libconfig-1.7.3-5.fc38.x86_64
$rpm -q perl
perl-5.36.1-497.fc38.x86_64

```


Which unit test is causing the segfault?

**Comment by yrutschle** [comment]
```
$ make test
"/usr/bin/perl" "-Iinc" -MExtUtils::Command::MM -e 'cp_nonempty' -- Libconfig.bs blib/arch/auto/Conf/Libconfig/Libconfig.bs 644
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'inc', 'blib/lib', 'blib/arch')" t/00-load.t t/01-init.t t/02-read.t t/03-write.t t/04-specread.t t/05-add.t t/06-delete.t t/07-memory.t t/08-exception.t t/09-complex.t t/10-mutiinit.t t/11-booltest.t t/11-mutinew.t t/12-includedir.t t/boilerplate.t t/pod-coverage.t t/pod-spell.t t/pod.t
t/00-load.t ........ 1/1 # Testing Conf::Libconfig 1.0, Perl 5.032001, /usr/bin/perl
t/00-load.t ........ ok   
t/01-init.t ........ ok   
t/02-read.t ........ ok     
t/03-write.t ....... ok   
t/04-specread.t .... Failed 9/10 subtests 
t/05-add.t ......... ok     
t/06-delete.t ...... ok     
t/07-memory.t ...... ok       
t/08-exception.t ... ok     
t/09-complex.t ..... ok   
t/10-mutiinit.t .... ok   
t/11-booltest.t .... ok     
t/11-mutinew.t ..... ok   
t/12-includedir.t .. ok    
t/boilerplate.t .... ok   
t/pod-coverage.t ... skipped: Test::Pod::Coverage 1.08 required for testing POD coverage
t/pod-spell.t ...... skipped: Test::Spelling required for testing
t/pod.t ............ skipped: Test::Pod 1.22 required for testing POD

Test Summary Report
-------------------
t/04-specread.t  (Wstat: 11 Tests: 1 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 10 tests but ran 1.
t/05-add.t       (Wstat: 0 Tests: 16 Failed: 0)
  TODO passed:   14
Files=18, Tests=31301,  5 wallclock secs ( 2.00 usr  0.17 sys +  4.56 cusr  0.49 csys =  7.22 CPU)
Result: FAIL
Failed 1/18 test programs. 0/31301 subtests failed.
make: *** [Makefile:944: test_dynamic] Error 255
```

This is with Debian's libconfig: 1.5-0.4.

I think I had more precise information but can't find it right now, I'll try and dig it up.
(I think however it had to do with array indexes)

**Comment by macportsraf** [comment]
The segfaulting test was t/04-specread.t on perl5.32. Your test run worked because you were using perl5.36. If you like, I could add some github actions to test it on multiple perl versions. It can definitely test from perl5.10 onwards, and possibly 5.8 as well.

**Comment by rafork** [comment]
The segfaulting test was t/04-specread.t on perl5.32. Your test run worked because you were using perl5.36. If you like, I could add some github actions to test it on multiple perl versions. It can definitely test from perl5.10 onwards, and possibly 5.8 as well.

**Comment by cnangel** [comment]
> The segfaulting test was t/04-specread.t on perl5.32. Your test run worked because you were using perl5.36. If you like, I could add some github actions to test it on multiple perl versions. It can definitely test from perl5.10 onwards, and possibly 5.8 as well.
I added you as a collaborator, you can try it. :)

**Comment by cnangel** [comment]
all cases have repaired. :)

**Comment by macportsraf** [comment]
Awesome! That's great. Many thanks.

**Comment by rafork** [comment]
Awesome! That's great. Many thanks.

**Comment by yrutschle** [comment]
:+1:  this installs fine, but my project still segfaults. I need to make sure my environment is clear of all the crap I injected when I investigated, and if it remains I'll post a minimum example as a new issue.

Thanks for the help, anyhow!
